### DISCUSSION TITLE
Add subcommands to LightningCLI

### DISCUSSION THREAD

**Discussion by vikigenius** [Lightning-AI/pytorch-lightning#13687] [Lightning Trainer API: Trainer, LightningModule, LightningDataModule] [answered]
I am not sure if this is possible and I am missing how to do this. But I couldn't figure out how to do it from the docs.

How can I add my own subcommands by extending LightningCLI ?

Say, my model code is part of an overall application I want to add a subcommand like `preprocess` for pre preparing data. Or `serve` for starting a dev webserver to serve the model etc ?

Is this possible with the current API? Are there any plans for it?

**Comment by mauvilsa** [comment, marked as answer]
It is possible. First you would subclass `Trainer` and add methods for each of the subcommands you want. These new methods should be usable independent of the CLI. Then you would need to subclass `LightningCLI`, in the `__init__` make your new trainer the default and override https://github.com/Lightning-AI/lightning/blob/d4c7f91fec6f4cd8edf5124a5cdfac08a07c4bde/src/pytorch_lightning/utilities/cli.py#L534-L543
